src/services/stringCompletions.ts

```diff
@@ -418,8 +424,19 @@ namespace ts.Completions.StringCompletions {
                     continue;
                 }
 
-                const foundFileName = includeExtensions || fileExtensionIs(filePath, Extension.Json) ? getBaseFileName(filePath) : removeFileExtension(getBaseFileName(filePath));
-                foundFiles.set(foundFileName, tryGetExtensionFromPath(filePath));
+                let foundFileName: string;
+                if (includeExtensionsOption === IncludeExtensionsOption.Exclude && !fileExtensionIs(filePath, Extension.Json)) {
+                    foundFileName = removeFileExtension(getBaseFileName(filePath));
+                    foundFiles.set(foundFileName, tryGetExtensionFromPath(filePath));
+                }
+                else if (includeExtensionsOption === IncludeExtensionsOption.ModuleSpecifierCompletion && fileExtensionIs(filePath, Extension.Ts)) {

```

Comment:
What we really need, and I’m not sure whether we have (but we might), is a function that maps input file extensions to emitted file extensions. So we would have

- .ts → .js
- .tsx → ? (I don’t know if this is always .js or if we might emit .jsx with `--jsx=preserve`?)
- .mts → .mjs (someday soon)
- .d.ts → `undefined` because it doesn’t get emitted
- .json → .json (only if `--resolveJsonModule`?)
- .js → .js

If we don’t have that function, we need to (a) figure out what the right thing for TSX is and implement it now, and (b) remember to update it for `.mts` and `.cts` if/when we add those.

Reply:
Emitter.ts has [`getOutputExtension`](https://github.com/microsoft/TypeScript/blob/b26f77a703b1f4e9e0cb07c02babbd708408beb0/src/compiler/emitter.ts#L113) which does approximately that, but it works on a sourceFile. Since we're just working working with the paths here we probably just want a new similar function that handle the cases we need.

Reply:
👍 in absence of something better, just leave comments on each function (`getOutputExtension`) and the new function saying to keep them in sync with each other.

Reply:
Found it: https://github.com/microsoft/TypeScript/blob/fad9122872a4df92050177edb07ee9cf3efd7e35/src/compiler/moduleSpecifiers.ts#L696

I would maybe just export that function so it doesn’t immediately cause merge conflicts with #44501.

Reply:
Actually, you might need to refactor it to return `undefined` instead of throwing on unsupported extensions, and move the debug assertions to its caller, or another wrapped version of it. But otherwise I think this function is exactly the thing we were looking for.